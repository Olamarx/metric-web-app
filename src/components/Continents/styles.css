.countries {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background-color: rgb(232, 35, 67);
}

.inPut-value {
  width: 70%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-bottom: 80px;
}

.search {
  width: 100%;
  height: 50px;
  outline: none;
  border: none;
}

.selectContinent {
  width: 100%;
  height: 50px;
  outline: none;
  border: none;
}

.countriesListing {
  list-style: none;
  display: grid;
  width: 100%;
  grid-template-columns: 1fr 1fr;
  gap: 0px;
  margin-right: 0;
}

.indCountry {
  width: 150px;
  height: 150px;
}

.indCountry a {
  color: white;
}

.twoImages {
  width: 200px;
  display: flex;
  gap: 10px;
}

.imageStyle {
  width: 70px;
  height: 70px;
}

.noCOA {
  width: 70px;
  height: 70px;
  margin-bottom: 30px;
}

.coatOfArm {
  width: 50px;
  height: 50px;
  border-radius: 30%;
}

.moreInfo {
  width: 200px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

@media screen and (min-width: 992px) {
  .imageStyle {
    width: 100px;
    height: 100px;
  }

  .moreInfo {
    width: 200px;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }


  .coatOfArm {
    width: 50px;
    height: 50px;
    border-radius: 30%;
  }

  .countriesListing {
    list-style: none;
    display: grid;
    width: 100%;
    grid-template-columns: 1fr 1fr;
    gap: 80px;
  }

  .countriesListing {
    grid-template-columns: 1fr 1fr 1fr 1fr;
  }
}
