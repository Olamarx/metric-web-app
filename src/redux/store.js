import { configureStore } from '@reduxjs/toolkit';
import countriesReducer from './Countries';
// import countryDetails, { loadCountry } from './MoreDetails';

const store = configureStore({
  reducer: {
    countriesReducer,
  },
});

export default store;
