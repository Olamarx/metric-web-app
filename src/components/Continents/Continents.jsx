import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import unidecode from 'unidecode';

import { getCountries } from '../../redux/continentCountries/continentCountries';
import { setContinent } from '../../redux/continent/continent';
import { setCountry } from '../../redux/country/country';
import store from '../../redux/store';
import Spinner from '../Loading/Loading';
import './styles.css';

function Continents() {
  const dispatch = useDispatch();
  const countries = useSelector((state) => state.countries);
  const continent = useSelector((state) => state.continent);
  const [searchCountries, setSearchCountries] = useState([]);
  const [query, setQuery] = useState('');
  const continents = ['Africa', 'Americas', 'Asia', 'Europe', 'Oceania'];

  const continentSelect = (e) => {
    setCountry(dispatch, {});
    setContinent(dispatch, e.target.value);
    getCountries(dispatch, store.getState, e.target.value);
  };

  const countrySelect = (country) => {
    setCountry(dispatch, country);
  };

  const filterSearchResult = (input) => {
    if (input === '') {
      return countries;
    }
    const result = countries.filter((country) => unidecode(`${country.name}`).toLowerCase().includes(unidecode(input)) || `${country.coords}`.includes(parseInt(input, 10)) || `${country.population}`.includes(parseInt(country, 10)));
    return result;
  };

  useEffect(() => {
    setCountry(dispatch, {});
    setSearchCountries(filterSearchResult(query));
  }, [countries, query]);

  useEffect(() => {
    getCountries(dispatch, store.getState, continent);
    setCountry(dispatch, {});
    setSearchCountries(filterSearchResult(query));
  }, [continent]);

  return (
    <div className="countries">
      <div className="inPut-value">
        <input type="text" className="search" value={query} placeholder="Search for Countries" onChange={(e) => setQuery(e.target.value)} />
        <select
          value={continent}
          onChange={(e) => continentSelect(e)}
          className="selectContinent"
        >
          {
            continents.map((continent) => (
              <option
                value={continent}
                key={continent}
              >
                {continent}
              </option>
            ))
        }
        </select>
      </div>

      <ul className="countriesListing">
        {
            searchCountries && searchCountries.map((country) => (
              <li className="indCountry" key={country.name}>
                <NavLink
                  to={`/continents/country/${unidecode(country.name.replace(/ /gi, '_'))}`}
                  className="link"
                  onClick={() => countrySelect(country)}
                >
                  <div className="twoImages">
                    <img className="imageStyle" src={country.flag} alt={`${country.name} flag`} />
                    {country.coatOfArm ? (
                      <img className="coatOfArm" src={country.coatOfArm} alt={`${country.name} Coat of arm`} />
                    ) : (
                      <div className="noCOA">
                        No Coat of Arm now,
                        check back later.
                      </div>
                    ) }

                  </div>
                  <div className="moreInfo">
                    <span>{country.name}</span>
                    <span>{`${country.population} population`}</span>
                  </div>
                </NavLink>
              </li>
            ))
        }
        {countries.length ? '' : <Spinner /> }
      </ul>
    </div>
  );
}

export default Continents;
