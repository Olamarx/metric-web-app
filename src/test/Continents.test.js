import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import Continents from '../components/Continents/Continents';
import store from '../redux/store';

describe('The indivial page of the app countires', () => {
  it('renders properly', async () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Provider store={store}>
          <Continents />
        </Provider>
      </MemoryRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
