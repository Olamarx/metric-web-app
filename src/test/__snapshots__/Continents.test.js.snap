// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The indivial page of the app countires renders properly 1`] = `
<div
  className="countries"
>
  <div
    className="inPut-value"
  >
    <input
      className="search"
      onChange={[Function]}
      placeholder="Search for Countries"
      type="text"
      value=""
    />
    <select
      className="selectContinent"
      onChange={[Function]}
      value="africa"
    >
      <option
        value="Africa"
      >
        Africa
      </option>
      <option
        value="Americas"
      >
        Americas
      </option>
      <option
        value="Asia"
      >
        Asia
      </option>
      <option
        value="Europe"
      >
        Europe
      </option>
      <option
        value="Oceania"
      >
        Oceania
      </option>
    </select>
  </div>
  <ul
    className="countriesListing"
  >
    <div
      className="ui active dimmer"
    >
      <div
        className="ui big text loader"
      >
        Loading...
      </div>
    </div>
  </ul>
</div>
`;
